_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+
+__auth.tsx__
+```tsx
+import backendAPI from "./base";
+
+let requestInterceptor: number;
+let responseInterceptor: number;
+
+export const clearInterceptors = () => {
+  backendAPI.interceptors.request.eject(requestInterceptor);
+  backendAPI.interceptors.response.eject(responseInterceptor);
+};
+
+export const setInterceptors = (accessToken: String) => {
+
+  clearInterceptors();
+
+  requestInterceptor = backendAPI.interceptors.request.use(
+    // @ts-expect-error
+    function (config) {
+      return {
+        ...config,
+        headers: {
+          ...config.headers,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      };
+    },
+    function (error) {
+      console.log("request interceptor error", error);
+      return Promise.reject(error);
+    }
+  );
+};
+```
+
+Add the file `src/hooks/useAccessToken.tsx` with the following content:
+
+__useAccessToken.tsx__
+```tsx
+import { setInterceptors } from "@/services/auth";
+import { useAuth0 } from "@auth0/auth0-react";
+import { useCallback, useState } from "react";
+
+export const useAccessToken = () => {
+  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
+  const [accessToken, setAccessToken] = useState("");
+
+  const saveAccessToken = useCallback(async () => {
+    if (isAuthenticated) {
+      try {
+        const tokenValue = await getAccessTokenSilently();
+        if (accessToken !== tokenValue) {
+          setInterceptors(tokenValue);
+          setAccessToken(tokenValue);
+        }
+      } catch (err) {
+        // Inactivity timeout
+        console.log("getAccessTokenSilently error", err);
+      }
+    }
+  }, [getAccessTokenSilently, isAuthenticated, accessToken]);
+
+  return {
+    saveAccessToken,
+  };
+};
+```
+
+The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the access token is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
+Create the `useAsyncWithToken` hook:
+
+__useAsyncWithToken.tsx__
+```tsx
+import { useAccessToken } from "./useAccessToken";
+import { useAsync } from "react-use-custom-hooks";
+
+export const useAsyncWithToken = <T, P, E = string>(
+  asyncOperation: () => Promise<T>, deps: any[]
+) => {
+  const { saveAccessToken } = useAccessToken();
+  const [ data, loading, error ] = useAsync(async () => {
+    await saveAccessToken();
+    return asyncOperation();
+  }, {},  deps);
+
+  return {
+    data,
+    loading,
+    error
+  };
+};
+```
+
+Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:
+
+__CompanyTableContainer.tsx__
+```tsx
+...
+const {
+  data: dataList,
+  loading: loadingList,
+  error: errorList,
+} = useAsyncWithToken(
+  () => CompanyApi.getCompanyList({ page: page - 1}),
+  [props.page]
+);
+
+const { data: dataCount } = useAsyncWithToken(
+  () => CompanyApi.getCompanyCount(),
+  []
+);
+...
```

Comment:
I had to add an import here:

```
import { useAsyncWithToken } from '@/hooks/useAsyncWithToken';
```

And this one is no longer needed:

```
import { useAsync } from 'react-use-custom-hooks';
```

Reply:
I am not sure how to explain the changes without showing the whole file again, I was trying to avoid repeating much the code. I re-formatted the block as diff.